### ISSUE TITLE
[Windows/VS]Building issues in Windows 10 related to linking

### ISSUE THREAD

**Issue by WinfredHuang** [facebookarchive/caffe2#1415]
My machine is Windows 10, CUDA 9.0, CuDNN 7.0, Visual Studio 2017 RTM, with latest version of source codes from GitHub. 
When building Caffe2, I proceeded according to the official installation guide, and built GFLAGS and GLOG with the help of CMake. I used /MD for both dependencies. Then, when I was building Caffe2, I left USE_STATIC_RUNTIME checked, BUILD_SHARED_LIBS unchecked, and CMake automatically used /MD for all projects. However, I encountered numerous linking errors. I changed options for several times, trying to mix up with different choices, but it still didn't work. 
I have followed the instructions of #1042 and modified all required files. I also think this is related to my previously submitted issue #1390 , although they may have different appearances. 
Before I give more details, I'd like to ask: Should I use /MT or /MD when building Caffe2 and its dependencies? 
@willyd I'm sorry to disturb you again, but I need help.

**Comment by willyd** [comment]
In windows development in general you should never mix `/MT` and `/MD`. 

Now you say that you build glog and gflags with `/MD` and have `USE_STATIC_RUNTIME` checked. This means that caffe2 will be compiled with `/MT` so you will have linker errors. Uncheck `USE_STATIC_RUNTIME` and `BUILD_SHARED_LIBS`. And this configuration should work.

You can have a look a the [caffe2 port](https://github.com/Microsoft/vcpkg/blob/master/ports/caffe2/portfile.cmake) in [vcpkg](https://github.com/Microsoft/vcpkg) and try it out. Normally, `vcpkg install caffe2:x64-windows` should build a static library with /MD and `vcpkg install caffe2:x64-windows-static` should build a static library with /MT. You can change CMake options in the portfile to include support for CUDA, OpenCV, etc.

Good luck!

**Comment by WinfredHuang** [comment]
Thank you. Now I have a little progress. 
I rebuilt Gflags and removed static libraries. Now I build Caffe2 and all dependencies with /MD. 
However I can't enable GPU support. Building on CPU context is OK. 
When I added "USE_CUDA", it reports: 
1. something like "fp16" is already defined in caffe2.lib, 
2. Many symbols defined but not referenced, and thus,
3. Numerous unresolved symbols (LNK2001). 
I'll continue testing CPU build to see if it's working. But I still need help to enable GPU mode. 
Again, my computer is NVIDIA GTX850M with CUDA 9.0 and cuDNN 7, under Windows 10 + Visual Studio 2017. 
I'll give more details on error messages later. 
Thank you again. 
**UPDATE:**
I built Caffe2 again with GPU support. I unchecked CUDA_USE_STATIC_CUDA_RUNTIME, and all other options are almost automatically generated by CMake. 
The following warnings and errors appear: (examples)
7>E:/caffe2/caffe2/operators/abs_op.cu(68): warning : function "caffe2::CAFFE_ANONYMOUS_VARIABLE_CUDAAbs" was declared but never referenced
...
7>C:/Program Files/gflags/include\gflags/gflags.h(226): warning : modifier is ignored on an enum specifier
...
7>E:/caffe2\caffe2/core/context_gpu.h(144): warning C4297: 'caffe2::CUDAContext::~CUDAContext': function assumed not to throw an exception but does
7>E:/caffe2\caffe2/core/context_gpu.h(144): note: destructor or deallocator has a (possibly implicit) non-throwing exception specification
...
7>E:/caffe2\caffe2/operators/conv_pool_op_base.h(519): warning : exception specification for virtual function "caffe2::ConvPoolOpBase<Context>::~ConvPoolOpBase [with Context=Context]" is incompatible with that of overridden function "caffe2::OperatorBase::~OperatorBase"
...
12>caffe2_gpu.lib(fully_connected_op_gpu.obj) : error LNK2005: "bool __cdecl caffe2::fp16_type<struct caffe2::__f16>(void)" (??$fp16_type@U__f16@caffe2@@@caffe2@@YA_NXZ) already defined in caffe2.lib(types.obj)
...
12>caffe2_gpu.lib(caffe2_gpu_generated_top_k.cu.obj) : error LNK2001: unresolved external symbol cudaDeviceGetAttribute
12>caffe2_gpu.lib(caffe2_gpu_generated_summarize_op.cu.obj) : error LNK2001: unresolved external symbol cudaDeviceGetAttribute
12>caffe2_gpu.lib(caffe2_gpu_generated_boolean_mask_ops.cu.obj) : error LNK2001: unresolved external symbol cudaPeekAtLastError
12>caffe2_gpu.lib(caffe2_gpu_generated_segment_reduction_op_gpu.cu.obj) : error LNK2001: unresolved external symbol cudaPeekAtLastError
12>caffe2_gpu.lib(caffe2_gpu_generated_perplexity_op.cu.obj) : error LNK2001: unresolved external symbol cudaPeekAtLastError
12>caffe2_gpu.lib(caffe2_gpu_generated_layer_norm_op.cu.obj) : error LNK2001: unresolved external symbol cudaPeekAtLastError
...
14>caffe2_gpu.lib(conv_transpose_op_cudnn.obj) : error LNK2001: unresolved external symbol cudnnCreate
14>caffe2_gpu.lib(conv_op_cudnn.obj) : error LNK2001: unresolved external symbol cudnnCreate
14>caffe2_gpu.lib(relu_op_cudnn.obj) : error LNK2001: unresolved external symbol cudnnCreate
14>caffe2_gpu.lib(recurrent_op_cudnn.obj) : error LNK2001: unresolved external symbol cudnnCreate
...
14>caffe2_gpu.lib(caffe2_gpu_generated_pool_op_cudnn.cu.obj) : error LNK2019: unresolved external symbol cudnnCreatePoolingDescriptor referenced in function "public: __cdecl caffe2::CuDNNPoolGradientOp::CuDNNPoolGradientOp(class caffe2::OperatorDef const &,class caffe2::Workspace *)" (??0CuDNNPoolGradientOp@caffe2@@QEAA@AEBVOperatorDef@1@PEAVWorkspace@1@@Z)
14>caffe2_gpu.lib(caffe2_gpu_generated_pool_op_cudnn.cu.obj) : error LNK2019: unresolved external symbol cudnnSetPooling2dDescriptor referenced in function "public: bool __cdecl caffe2::CuDNNPoolGradientOp::DoRunWithType<float,float>(void)" (??$DoRunWithType@MM@CuDNNPoolGradientOp@caffe2@@QEAA_NXZ)
14>caffe2_gpu.lib(caffe2_gpu_generated_pool_op_cudnn.cu.obj) : error LNK2019: unresolved external symbol cudnnSetPoolingNdDescriptor referenced in function "public: bool __cdecl caffe2::CuDNNPoolGradientOp::DoRunWithType<float,float>(void)" (??$DoRunWithType@MM@CuDNNPoolGradientOp@caffe2@@QEAA_NXZ)
14>caffe2_gpu.lib(caffe2_gpu_generated_pool_op_cudnn.cu.obj) : error LNK2019: unresolved external symbol cudnnDestroyPoolingDescriptor referenced in function "public: virtual __cdecl caffe2::CuDNNPoolGradientOp::~CuDNNPoolGradientOp(void)" (??1CuDNNPoolGradientOp@caffe2@@UEAA@XZ)
14>caffe2_gpu.lib(caffe2_gpu_generated_pool_op_cudnn.cu.obj) : error LNK2019: unresolved external symbol cudnnPoolingForward referenced in function "public: bool __cdecl caffe2::CuDNNPoolOp::DoRunWithType<float,float>(void)" (??$DoRunWithType@MM@CuDNNPoolOp@caffe2@@QEAA_NXZ)
14>caffe2_gpu.lib(caffe2_gpu_generated_pool_op_cudnn.cu.obj) : error LNK2019: unresolved external symbol cudnnPoolingBackward referenced in function "public: bool __cdecl caffe2::CuDNNPoolGradientOp::DoRunWithType<float,float>(void)" (??$DoRunWithType@MM@CuDNNPoolGradientOp@caffe2@@QEAA_NXZ)
...
And below is my CMake generation log:

Selecting Windows SDK version 10.0.14393.0 to target Windows 10.0.16299.
Could NOT find Protobuf (missing: Protobuf_DIR)
Use custom protobuf build.
Using protobuf compiler $<TARGET_FILE:protoc>
Could NOT find Git (missing: GIT_EXECUTABLE) 
CMake Warning at cmake/Dependencies.cmake:33 (message):
  Cannot find git, so Caffe2 won't have any git build info available
Call Stack (most recent call first):
  CMakeLists.txt:79 (include)


The BLAS backend of choice:Eigen
Found gflags  (include: C:/Program Files/gflags/include, library: gflags_shared)
Found glog    (include: E:/google/glog/common;E:/google/glog/src;E:/google/glog/src/windows, library: glog::glog)
Found lmdb    (include: D:/Anaconda2/Library/include, library: lmdb)
Could NOT find LevelDB (missing: LevelDB_INCLUDE LevelDB_LIBRARY) 
Could NOT find Snappy (missing: Snappy_INCLUDE_DIR Snappy_LIBRARIES) 
CMake Warning at cmake/Dependencies.cmake:163 (message):
  Not compiling with LevelDB.  Suppress this warning with -DUSE_LEVELDB=OFF
Call Stack (most recent call first):
  CMakeLists.txt:79 (include)


OpenCV found (E:/opencv/build/x64/vc14/lib)
CMake Warning at cmake/Dependencies.cmake:239 (find_package):
  By not providing "FindEigen3.cmake" in CMAKE_MODULE_PATH this project has
  asked CMake to find a package configuration file provided by "Eigen3", but
  CMake did not find one.

  Could not find a package configuration file provided by "Eigen3" with any
  of the following names:

    Eigen3Config.cmake
    eigen3-config.cmake

  Add the installation prefix of "Eigen3" to CMAKE_PREFIX_PATH or set
  "Eigen3_DIR" to a directory containing one of the above files.  If "Eigen3"
  provides a separate development package or SDK, be sure it has been
  installed.
Call Stack (most recent call first):
  CMakeLists.txt:79 (include)


Did not find system Eigen. Using third party subdirectory.
NumPy ver. 1.13.1 found (include: D:/Anaconda2/lib/site-packages/numpy/core/include)
Could NOT find MPI_C (missing: MPI_C_COMPILER) 
Could NOT find MPI_CXX (missing: MPI_CXX_COMPILER) 
Could NOT find MPI (missing: MPI_C_FOUND MPI_CXX_FOUND) 
CMake Warning at cmake/Dependencies.cmake:299 (message):
  Not compiling with MPI.  Suppress this warning with -DUSE_MPI=OFF
Call Stack (most recent call first):
  CMakeLists.txt:79 (include)


Adding -openmp
CUDA detected: 9.0
Added CUDA NVCC flags for: sm_30 sm_35 sm_50 sm_52 sm_60 sm_61 sm_70
Found libcuda: D:/NVIDIA/CUDA/lib/x64/cuda.lib
Found libnvrtc: D:/NVIDIA/CUDA/lib/x64/nvrtc.lib
Found cuDNN: v7.0.3  (include: D:/NVIDIA/CUDA/include, library: D:/NVIDIA/CUDA/lib/x64/cudnn.lib)
CMake Warning at cmake/Dependencies.cmake:357 (message):
  NCCL is currently only supported under Linux.
Call Stack (most recent call first):
  CMakeLists.txt:79 (include)


Need to define long as a separate typeid.
Current compiler supports avx2 extention. Will build perfkernels.
CMake Warning (dev) at cmake/MiscCheck.cmake:75 (if):
  Policy CMP0054 is not set: Only interpret if() arguments as variables or
  keywords when unquoted.  Run "cmake --help-policy CMP0054" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  Quoted variables like "MSVC" will no longer be dereferenced when the policy
  is set to NEW.  Since the policy is not set the OLD behavior will be used.
Call Stack (most recent call first):
  CMakeLists.txt:82 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.

NCCL operators skipped due to no CUDA support
Including image processing operators
Excluding video processing operators due to no opencv
Excluding mkl operators as we are not using mkl
MPI operators skipped due to no MPI support
Include Observer library
CMake Warning (dev) at cmake/Utils.cmake:176 (if):
  Policy CMP0054 is not set: Only interpret if() arguments as variables or
  keywords when unquoted.  Run "cmake --help-policy CMP0054" for policy
  details.  Use the cmake_policy command to set the policy and suppress this
  warning.

  Quoted variables like "MSVC" will no longer be dereferenced when the policy
  is set to NEW.  Since the policy is not set the OLD behavior will be used.
Call Stack (most recent call first):
  cmake/Utils.cmake:193 (caffe_add_whole_archive_flag)
  caffe2/CMakeLists.txt:106 (caffe_add_linker_flag)
This warning is for project developers.  Use -Wno-dev to suppress it.

Automatically generating missing __init__.py files.

******** Summary ********
General:
  Git version           : 
  System                : Windows
  C++ compiler          : D:/Microsoft Visual Studio/2017/Enterprise/VC/Tools/MSVC/14.10.25017/bin/HostX86/x64/cl.exe
  C++ compiler version  : 19.10.25017.0
  Runtime               : dynamic (/MD)
  Protobuf compiler     : $<TARGET_FILE:protoc>
  CXX flags             : /DWIN32 /D_WINDOWS /W3 /GR /EHsc -openmp /MP /bigobj
  Build type            : Release
  Compile definitions   : 

  BUILD_BINARY          : ON
  BUILD_PYTHON          : ON
    Python version      : 2.7.13
    Python library      : D:/Anaconda2/libs/python27.lib
  BUILD_SHARED_LIBS     : OFF
  BUILD_TEST            : OFF
  USE_ATEN              : OFF
  USE_ASAN              : OFF
  USE_CUDA              : ON
    CUDA version        : 9.0
    CuDNN version       : 7.0.3
  USE_EIGEN_FOR_BLAS    : 1
  USE_FFMPEG            : OFF
  USE_GFLAGS            : ON
  USE_GLOG              : ON
  USE_GLOO              : OFF
  USE_LEVELDB           : OFF
  USE_LITE_PROTO        : OFF
  USE_LMDB              : ON
    LMDB version        : 0.9.21
  USE_METAL             : OFF
  USE_MKL               : 
  USE_MOBILE_OPENGL     : OFF
  USE_MPI               : OFF
  USE_NCCL              : OFF
  USE_NERVANA_GPU       : OFF
  USE_NNPACK            : OFF
  USE_OBSERVERS         : ON
  USE_OPENCV            : ON
    OpenCV version      : 3.3.0
  USE_OPENMP            : ON
  USE_REDIS             : OFF
  USE_ROCKSDB           : OFF
  USE_THREADS           : ON
  USE_ZMQ               : OFF
CMake Warning at CMakeLists.txt:152 (message):
  Generated cmake files are only fully available if one builds with system
  glog, gflags and protobuf.


Configuring done
Generating done
